### PR TITLE
 Unmanned vehicles can be controlled by AI + added to req

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Unmanned vehicles can be bought from req, there also two droids already on ship (one scout and one combat). 
Relayed speech added to remote control component
Ai can natively control these vehicles
Rebalance them a bit

## Changelog
:cl:
add: Unmanned vehicles and remote control added to req
add: Ai can control unmanned vehicles
add: Relayed speech added to droids
/:cl:

### CODE REVIEW

**Comment by ilikesaying** [comment]
no way

**Comment by vilereaver41** [comment]
What does combat one do fighting wise, sounds fun. Now hivemind needs some

**Review by TiviPlus — CHANGES_REQUESTED** [review]
You need to make the droids bump attackable before players can use them

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 1]
```diff
@@ -45,6 +45,8 @@
 	var/datum/trackable/track
 	///Selected order to give to marine
 	var/datum/action/innate/order/current_order
+	/// The current controlled vehicle
+	var/obj/vehicle/unmanned/vehicle
```
Comment (TiviPlus) [inline code/modules/mob/living/silicon/ai/ai.dm:49]: Dont hard ref it on the AI like this

File: code/_globalvars/lists/objects.dm [hunk 2]
```diff
@@ -45,6 +45,8 @@ GLOBAL_LIST_EMPTY(xeno_resin_door_turfs)
 
 GLOBAL_LIST_EMPTY(shuttle_controls_list)
 
+GLOBAL_LIST_EMPTY(vehicles)
+
```
Comment (TiviPlus) [inline code/_globalvars/lists/objects.dm:49]: this is unmanned vehicles and not all vehicles

File: code/modules/vehicles/unmanned/unmanned_vehicle.dm [hunk 3]
```diff
@@ -29,12 +30,33 @@
...
+	var/static/serial = 1
+	/// If the vehicle should spawn with a weapon allready installed
+	var/obj/item/uav_turret/spawn_equipped_type = null
 	COOLDOWN_DECLARE(fire_cooldown)
 
 /obj/vehicle/unmanned/Initialize()
 	. = ..()
 	current_rounds = max_rounds
 	ammo = GLOB.ammo_list[ammo]
+	name = name + " " + num2text(serial)
```
Comment (TiviPlus) [inline code/modules/vehicles/unmanned/unmanned_vehicle.dm:45]: += exists

File: code/__DEFINES/equipment.dm [hunk 4]
```diff
@@ -28,6 +28,8 @@
 #define PREVENT_CONTENTS_EXPLOSION	(1<<13)
 /// was this spawned by an admin? used for stat tracking stuff.
 #define ADMIN_SPAWNED			(1<<14)
+/// Can this atom be bumped into
+#define BUMPABLE (1<<15)
```
Comment (TiviPlus) [inline code/__DEFINES/equipment.dm:32]: Bump_attackable?

File: code/modules/vehicles/unmanned/unmanned_vehicle.dm [hunk 5]
```diff
@@ -57,8 +95,48 @@
 
 /obj/vehicle/unmanned/attackby(obj/item/I, mob/user, params)
 	. = ..()
-	if(!istype(I, /obj/item/uav_turret) && !istype(I, /obj/item/explosive/plastique))
+	if(istype(I, /obj/item/tool/wrench))
+		return remove_turret(user)
```
Comment (TiviPlus) [inline code/modules/vehicles/unmanned/unmanned_vehicle.dm:99]: wrench_act

File: code/modules/vehicles/unmanned/unmanned_vehicle.dm [hunk 6]
```diff
@@ -57,8 +95,48 @@
...
+	if(istype(I, /obj/item/uav_turret) || istype(I, /obj/item/explosive/plastique))
+		return equip_turret(I, user)
```
Comment (TiviPlus) [inline code/modules/vehicles/unmanned/unmanned_vehicle.dm:101]: iirc plastique has a isthing but I may be wrong

Comment (BraveMole) [reply code/modules/vehicles/unmanned/unmanned_vehicle.dm:103]: nop

**Comment by Wayland-Smithy** [comment]
Thought it was cool until a minitank turned out to be completely unmovable by xenos? I could not as Crusher move it on any intent or with any ability, even bump attack. Please reply if this is intended or not.

Unsure if it applies to the bipedal models too, but apparently they also are not targeted by xeno turrets or affected by sticky resin.

**Comment by jroinc** [comment]
Droids are REALLY fast and unaffected by most slowdowns or ayy abilities.

A round where all ungas had one led to almost all of the xenos dying within one minute of shutters, because the droids were pacing runners.

Even once HP was dropped to 200 from standard 300, and most ayy hp was doubled, drones were still far more effective than normal ungas.

Drop the speed slightly? Otherwise it's probably fine.

**Comment by BraveMole** [comment]
> Thought it was cool until a minitank turned out to be completely unmovable by xenos? I could not as Crusher move it on any intent or with any ability, even bump attack. Please reply if this is intended or not.

Yes I am aware of this, this is mostly code limitation. I don't dislike the idea of unmovable vehicle though, so I don't intend to change that

> Unsure if it applies to the bipedal models too, but apparently they also are not targeted by xeno turrets or affected by sticky resin.

Adding anykind of slowdown would mean a lot of code. I'll fix the xeno turrets thing, and probably slow down a bit everything.

**Comment by BraveMole** [comment]
> Droids are REALLY fast and unaffected by most slowdowns or ayy abilities.

Yeah, nerfing the speed. They'll keep being unaffected by most xeno ability, I'll maybe tweak for critical ones (shrike fling maybe?)

> A round where all ungas had one led to almost all of the xenos dying within one minute of shutters, because the droids were pacing runners.

Ew

> Even once HP was dropped to 200 from standard 300, and most ayy hp was doubled, drones were still far more effective than normal ungas.

Them being effective is not that such of an issue IMO, since they are behind req.

> Drop the speed slightly? Otherwise it's probably fine.
> Also add in a one-second cooldown for the activation/deactivation sound, it's very spammable.

Yeah ok

**Comment by SilencedMP5A5** [comment]
please for the love of god get rid of the fucking sounds they make i hate the sounds so much kill it kill it kill it

**Comment by jroinc** [comment]
Now that this is unTM'd, I have one other additional concern. 

There's no real admin logging associated with these, so someone who bought one and shot up the FOB would be very difficult to track down. Impossible for most admins. Can you add some sort of FF logging to them, and some sort of way to tell which ckey is driving them with that log?

**Comment by BraveMole** [comment]
> Now that this is unTM'd, I have one other additional concern.
> 
> There's no real admin logging associated with these, so someone who bought one and shot up the FOB would be very difficult to track down. Impossible for most admins. Can you add some sort of FF logging to them, and some sort of way to tell which ckey is driving them with that log?

Yeah working on it

**Comment by Wayland-Smithy** [comment]
> I don't dislike the idea of unmovable vehicle though, so I don't intend to change that

That's pretty balance breaking.

File: code/datums/components/remote_control.dm [hunk 7]
```diff
@@ -5,53 +5,86 @@
...
 	SIGNAL_HANDLER
 	switch(type)
-		if(TURRET_TYPE_HEAVY, TURRET_TYPE_LIGHT, TURRET_TYPE_DROIDLASER)
-			click_proc = CALLBACK(src, .proc/uv_handle_click)
-		if(TURRET_TYPE_EXPLOSIVE)
-			click_proc = CALLBACK(src, .proc/uv_handle_click_explosive)
+		if(/obj/item/uav_turret, /obj/item/uav_turret/heavy, /obj/item/uav_turret/droid)
+			left_click_proc = CALLBACK(src, .proc/uv_handle_click)
+		if(/obj/item/explosive/plastique)
+			left_click_proc = CALLBACK(src, .proc/uv_handle_click_explosive)
```
Comment (TiviPlus) [inline code/datums/components/remote_control.dm:63]: why

File: code/modules/xenomorph/xeno_structures.dm [hunk 8]
```diff
@@ -440,6 +442,9 @@
 		if(nearby_xeno.stat == DEAD)
 			continue
 		potential_hostiles += nearby_xeno
+	for(var/obj/vehicle/unmanned/vehicle AS in GLOB.unmanned_vehicles)
+		if(get_dist(vehicle, src) <= range)
+			potential_hostiles += vehicle
```
Comment (TiviPlus) [inline code/modules/xenomorph/xeno_structures.dm:447]: But why

File: code/datums/components/remote_control.dm (hunk 7 above)
Comment (BraveMole) [reply code/datums/components/remote_control.dm:63]: This is your code?

File: code/modules/xenomorph/xeno_structures.dm (hunk 8 above)
Comment (BraveMole) [reply code/modules/xenomorph/xeno_structures.dm:447]: So xeno turrets can attack unmanned vehicle?

File: code/datums/components/remote_control.dm (hunk 7 above)
Comment (TiviPlus) [reply code/datums/components/remote_control.dm:63]: Why did you change it to a strict type check from a var check

Comment (TiviPlus) [reply code/datums/components/remote_control.dm:63]: Change it back strict type chacks like this are bad

**Comment by Derpguy3** [comment]
Some feedback regarding a few things.

AI Controlling
- The AI is unable to fully see things around the controlled drone, due to its vision being limited to cameras or marine helmet cams.
- The camera view is occasionally moved from the drone when firing, as a double click causes the camera view to either move or track a mob.

Drone feedback

- The scout droid's cloak ability is difficult to use for its scouting purposes due to the short period of cloaked time.
- The droid energy gun has no way to reload besides discarding for a new one.

**Comment by BraveMole** [comment]
> AI Controlling
> 
> * The AI is unable to fully see things around the controlled drone, due to its vision being limited to cameras or marine helmet cams.
> * The camera view is occasionally moved from the drone when firing, as a double click causes the camera view to either move or track a mob.

Both fixed

> Drone feedback
> * The scout droid's cloak ability is difficult to use for its scouting purposes due to the short period of cloaked time.

Yeah scout droid is kinda useless right now, i'll rebalance it

> * The droid energy gun has no way to reload besides discarding for a new one.

Intended. Don't ask me why, i thought it was cool

**Pull request closed without merging**